WIP: Implement cart skeleton

nielslange opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR updates the cart skeleton to introduce a unified loading state. The new skeleton closely matches the layout of the rendered Cart block and appears instantly on page load, thanks to its server-rendered HTML markup. This ensures users see a consistent layout while cart data is loading.

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOPLUG-3467
Closes #56442

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

### Desktop

<table>
<tr>
<td valign="top">Before:
<br><br>
<img width="1512" alt="Screenshot 2025-04-14 at 11 57 15" src="https://github.com/user-attachments/assets/7a5bf4da-9f62-4717-9761-b24a3a52663c" />
</td>
<td valign="top">After:
<br><br>
<img width="1512" alt="Screenshot 2025-04-14 at 11 45 45" src="https://github.com/user-attachments/assets/e2a1ae96-0fd3-4076-be66-18fd4069f210" />
</td>
<td valign="top">Rendered block:
<br><br>
<img width="1512" alt="Screenshot 2025-04-14 at 12 50 26" src="https://github.com/user-attachments/assets/28cca1d0-5bce-4b72-aadc-ffaaef311f0e" />
</td>
</tr>
</table>

### Mobile

<table>
<tr>
<td valign="top">Before:
<br><br>
<img width="346" alt="Screenshot 2025-04-14 at 12 28 27" src="https://github.com/user-attachments/assets/197d5a18-9869-4fd4-9194-fb666ff0500b" />
</td>
<td valign="top">After:
<br><br>
<img width="290" alt="Screenshot 2025-04-14 at 13 13 32" src="https://github.com/user-attachments/assets/c1fc716a-13d3-40ee-8c86-eb77905ba3b7" />
</td>
<td valign="top">Rendered block:
<br><br>
<img width="287" alt="Screenshot 2025-04-14 at 13 14 31" src="https://github.com/user-attachments/assets/45dacf0c-c65c-4632-a854-6cf6790756db" />
</td>
</tr>
</table>


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add a product to the cart.
2. Go to the cart page, using the Cart block.
3. Open the inspector.
4. Throttle down the network speed, e.g.  to `3G`.
5. Reload the cart page.
6. Verify that the new skeleton loader is visible.
7. Test the cart skeleton in different browsers and with different resultions.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

nielslange committed:
Refactor Skeleton Component and Add New Patterns

- Updated the Skeleton component to accept width, height, and borderRadius props, replacing the numberOfLines prop.
- Introduced new skeleton patterns: CartExpressPaymentsSkeleton, CheckoutExpressPaymentsSkeleton, and ProductNoticeSkeleton for improved loading states.
- Enhanced styling for skeleton elements to support a shimmer effect and better visual representation in Storybook.
- Updated references in AddToCartWithOptions and AddToCartForm blocks to utilize the new ProductNoticeSkeleton.

nielslange committed:
Enhance Skeleton Component with New Layouts and Patterns

- Added new layouts for Cart and Checkout skeletons to improve loading states.
- Introduced additional skeleton patterns: CartLineItemsSkeleton, CartOrderSummarySkeleton, CheckoutContactSkeleton, CheckoutDeliverySkeleton, CheckoutPaymentSkeleton, and CheckoutShippingSkeleton.
- Updated the Skeleton component to support maxWidth prop and adjusted default height.
- Refined styles for skeleton elements to enhance visual representation in Storybook.
- Updated references in AddToCartWithOptions and AddToCartForm blocks to utilize the new ProductNoticeSkeleton.

nielslange committed:
Add changelog entry and address lint issues

nielslange committed:
Address lint issues by temporarily disabling stylelint for the affected sections.

nielslange committed:
Create CartFullSkeleton, CheckoutFullSkeleton and CheckoutOrderSummaryMobileSkeleton

nielslange committed:
Merge branch 'trunk' into add/56480-skeleton-component-system

nielslange committed:
Refactor skeleton component imports and story names for consistency

- Updated imports in cart, checkout, and checkout-full layouts to remove unnecessary component suffixes.
- Renamed story exports in base and patterns stories for clarity and uniformity, enhancing readability.
- Adjusted height and width properties in stories to improve visual representation of skeleton components.

nielslange committed:
Merge branch 'trunk' into add/56480-skeleton-component-system

nielslange committed:
Merge branch 'trunk' into add/56480-skeleton-component-system

nielslange committed:
Merge branch 'trunk' into add/56480-skeleton-component-system

nielslange committed:
Merge branch 'trunk' into add/56480-skeleton-component-system

nielslange committed:
Add CheckoutOrderSummaryMobileSkeleton to CheckoutSkeleton

nielslange committed:
Replace ProductNoticeSkeleton with ProductShortDescriptionSkeleton

nielslange committed:
WIP: Add CartSkeleton component and integrate loading state handling

- Introduced a new CartSkeleton component for displaying a loading state in the cart block.
- Updated frontend.js to render the CartSkeleton and remove it once the cart has finished loading.
- Modified Cart.php to include a div for the CartSkeleton in the block's output.
- Adjusted style.scss to clean up unused styles related to the cart block.

nielslange committed:
Enhance Cart block with CartSkeleton component for improved loading state

- Added a new CartSkeleton class to generate skeleton HTML for the cart block.
- Integrated CartSkeleton into the Cart block's output to display a loading state.
- Updated styles in style.scss to support the new skeleton layout and improve visual presentation.
- Ensured responsive design adjustments for the cart block's sidebar layout.

nielslange committed:
Remove unused CartSkeleton component and its import from frontend.js

- Deleted the cart-skeleton.tsx file as it is no longer needed.
- Removed the import statement for renderCartSkeleton in frontend.js to clean up the codebase.

nielslange committed:
Adjust skeleton color

nielslange committed:
Refactor Cart block loading state handling

- Removed the LoadingMask component from the Cart block to streamline the rendering process.
- Enhanced frontend.js to add a loading class to the cart element during loading, improving user experience.
- Updated style.scss to include new styles for the loading state and skeleton elements, ensuring a smoother visual transition during loading.

nielslange committed:
Update Cart block styles and enhance skeleton loading elements

- Removed unused loading styles from style.scss to streamline the cart block's appearance.
- Updated CartSkeleton.php to include skeleton elements for the cart header, improving the visual representation during loading.
- Adjusted styles for better alignment and spacing in the cart block's skeleton layout.

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Fix lint errors

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Implement strict typing in CartSkeleton.php

nielslange committed:
Refactor CartSkeleton.php to remove unnecessary aria-hidden attributes from skeleton elements.

nielslange commented on plugins/woocommerce/client/blocks/assets/js/blocks/cart/frontend.js:
```diff
@@ -27,7 +31,26 @@ const getProps = ( el ) => {
+
+			if ( cartElement && skeletonElement ) {
+				cartElement.classList.add( 'is-loading' );
+				setTimeout( () => {
+					skeletonElement.remove();
+					cartElement.classList.remove( 'is-loading' );
+				}, 1000 );
```
The timeout is intentionally hard-coded to prevent a flicker and smoothen the UX. Without it, there's a brief flash. These screen recordings show the difference. That said, I'm keen to hear the thoughts of @pierorocca and @woocommerce/rubik on this approach.

## Without `setTimeout`

https://github.com/user-attachments/assets/117f1a05-adff-4745-9163-41ceaefa6666

## With `setTimeout`

https://github.com/user-attachments/assets/a0707440-d8f5-4d7d-8912-7351942c9868

tjcafferkey replied:
I definitely don't think we should be adding in setTimeout's. We should be delivering content to users as quickly as possible rather than delaying it. Also, what if hypothetically the cart already took 5 seconds to load (which is obviously too long in itself)? In this scenario we'd be unnecessarily creating a 6 second overall delay.

nielslange replied:
You're right, @tjcafferkey. I added the timeout to avoid a flicker when the Cart block loads quickly (see video above), but I agree it’s not ideal to delay content. I’ve removed it now. I would appreciate a second look at the PR and your thoughts on how the flicker feels without the timeout.

nielslange committed:
Add changelog entry

tjcafferkey commented on plugins/woocommerce/src/Blocks/Skeleton/CartSkeleton.php:
```diff
@@ -0,0 +1,87 @@
+<?php
```
I haven't dived deeply into the implementation yet but I am curious as to why `CartSkeleton.php` is so separate and not part of the `Cart.php` class? It seems really specific to the block and having it separate feels like it creates some fragmentation. What is the benefit to your approach here?

nielslange replied:
I created `CartSkeleton.php` to keep the `render()` function in `Cart.php` readable by moving the HTML markup out. I initially planned to use the [skeleton components](https://github.com/woocommerce/woocommerce/issues/56480), but due to the timing of when the Cart block renders, they’re not usable here. Instead, I want to reuse the HTML markup for Storybook stories, which is another reason for this separation.

tjcafferkey commented:
I don't know enough about the cart and checkout code yet to know if there is a reason us not doing the following, but what if each inner block was responsible for loading its own skeleton element rather than us writing all of this upfront as a HTML block. This way if an inner block is present it will render its own skeleton placeholder, more accurately representing the loaded state.

nielslange commented:
> I don't know enough about the cart and checkout code yet to know if there is a reason us not doing the following, but what if each inner block was responsible for loading its own skeleton element rather than us writing all of this HTML upfront in a single block of code. This way if an inner block is present it will render its own skeleton placeholder, more accurately representing the loaded state.

In theory, that sounds like a great approach. In practice, though, it won’t work due to how the Cart block renders. The HTML markup is available immediately on page load, but inner block skeletons would appear much later. That’s because we first enqueue JS assets, then begin hydration. Only after that does `useStoreCart` initialise and the `cartIsLoading` state become available. If we relied on inner block skeletons, all of that would need to happen before anything is shown.

nielslange committed:
Optimize cart loading behavior by removing the loading class and skeleton element immediately upon cart data availability in frontend.js.

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Enhance cart block styles with grid layout for cart items

- Introduced a grid layout for cart items to improve visual structure and alignment.
- Removed the cart items header for a cleaner appearance.
- Adjusted padding and margins for better spacing within the cart block.

nielslange committed:
Merge branch 'update/56442-implement-cart-skeleton' of github.com:woocommerce/woocommerce into update/56442-implement-cart-skeleton

tjcafferkey commented:
>In theory, that sounds like a great approach. In practice, though, it won’t work due to how the Cart block renders. The HTML markup is available immediately on page load, but inner block skeletons would appear much later. That’s because we first enqueue JS assets, then begin hydration. Only after that does useStoreCart initialise and the cartIsLoading state become available. If we relied on inner block skeletons, all of that would need to happen before anything is shown.

Sorry, perhaps I wasn't really clear in what I meant. We render the skeleton server-side for each inner block and when the JS is downloaded it will mount in its place, replacing the skeleton.

samueljseay commented:
@nielslange One thing I was a bit confused about is that it seems you have implemented a whole component system here: https://github.com/woocommerce/woocommerce/pull/56719 to do Skeleton components.

And then in this PR you statically render some HTML for the skeleton. As far as I can tell it's supposed to be part of the same system. 

1. If that's not the case, what is the purpose of the previous PR can you explain what it was for?
2. If it's the case that actually all you needed was some static HTML and some styles (which seems fair to me, the component system looks way overkill for this afaict) can you please remove those components you introduced as part of this PR

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Refactor cart block styles and remove unused CartSkeleton class

- Updated SCSS for cart line items and cart blocks to improve layout and spacing.
- Enhanced loading states for empty cart and order summary components.
- Removed the CartSkeleton class as its functionality is now integrated directly into the cart block.
- Added skeleton loading elements for better user experience during data fetching.

nielslange committed:
Remove unused skeleton components for cart and checkout layouts

- Deleted CartSkeleton, CartFullSkeleton, CheckoutSkeleton, and CheckoutFullSkeleton components along with their associated styles.
- Cleaned up related patterns and stories to streamline the codebase.
- This refactor improves maintainability by removing redundant code and focusing on essential components.

nielslange committed:
Merge branch 'update/56442-implement-cart-skeleton' of github.com:woocommerce/woocommerce into update/56442-implement-cart-skeleton

nielslange committed:
Enhance cart block styles and improve layout

- Updated SCSS for cart line items and cart components to refine padding and spacing.
- Added new styles for cart order summary heading and totals blocks.
- Removed unused skeleton styles to streamline the codebase and improve maintainability.

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton

nielslange committed:
Fix failing e2e tests

nielslange committed:
Merge branch 'trunk' into update/56442-implement-cart-skeleton
